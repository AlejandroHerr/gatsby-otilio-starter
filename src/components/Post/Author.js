// @flow
import React from 'react';
import Img from 'gatsby-image';

import type { AuthorType } from '../../types/post';

import ByteIcon from '../icons/ByteIcon';
import SocialLink from './SocialLink';

import styles from './Post.module.scss';

const Author = ({ author }: {author: AuthorType}) => (
  <aside className={styles.post_author}>
    <figure className={`${styles.post_author__avatar} ${styles.avatar}`}>
      <Img sizes={author.avatar.childImageSharp.sizes} />
    </figure>
    <div className={styles.post_author__bio}>
      <h4 className={styles.post_author__name}>{author.name}</h4>
      <p className={styles.post_author__about}>{author.about}</p>
      <span className={styles.post_author__location}>
        <ByteIcon className={styles.post_author__location_icon} icon="location" /> {author.location}
      </span>
      {author.social.map(({ service, url }) => (
        <SocialLink key={service} service={service} url={url} />
      ))}
    </div>
  </aside>
);

export default Author;
