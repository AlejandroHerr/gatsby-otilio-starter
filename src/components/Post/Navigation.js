// @flow
import React from 'react';

import type { PreviewType } from '../../types/post';

import NavigationLink from './NavigationLink';

import styles from './Post.module.scss';

const Navigation = ({ next, prev }: {next ?: PreviewType, prev?: PreviewType}) => (
  <aside className={styles.post_navigation}>
    {next &&
      <NavigationLink
        className={styles.post_navigation__next}
        icon="chevron_left"
        {...next}
      />
    }
    {prev &&
      <NavigationLink
        className={styles.post_navigation__prev}
        icon="chevron_right"
        {...prev}
      />
    }
    <div className={styles.post_navigation__clear} />
    <div className={styles.post_navigation__separator} />
  </aside>
);

Navigation.defaultProps = {
  next: null,
  prev: null,
};

export default Navigation;
