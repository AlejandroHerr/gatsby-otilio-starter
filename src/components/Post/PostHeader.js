import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Link from 'gatsby-link';

import ClockIcon from '../icons/ClockIcon';
import CoverImage from '../CoverImage';

import { imageSharp } from './propTypes';

import styles from './PostHeader.module.scss';

const getDisplayDate = date => date.toLocaleDateString('en-us', { day: 'numeric', month: 'short', year: 'numeric' });

const PostHeader = ({
  author, date, image, title, timeToRead,
}) => {
  const postDate = new Date(date);
  return (
    <header className={classnames(styles.post_header, { [styles['post_header--has_image']]: !!image })}>
      <div className={styles.post_header__inner}>
        <h1 className={styles.post_header__title}>
          {title}
        </h1>
        <span className={styles.post_header__body}>
          <Link className={styles.post_header__author} to="/">{author}</Link>
          {' | '}
          <span><ClockIcon /> {timeToRead}</span>
          {' | '}
          <time dateTime={postDate.toISOString()}>{getDisplayDate(postDate)}</time>
        </span>
        {image && (
        <CoverImage
          className={styles.post_header__image}
          shadowClassName={styles.post_header__image_shadow}
          image={image}
        />
      )}
      </div>
    </header>
  );
};


PostHeader.defaultProps = {
  image: null,
};

PostHeader.propTypes = {
  author: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  image: imageSharp,
  title: PropTypes.string.isRequired,
  timeToRead: PropTypes.number.isRequired,
};

export default PostHeader;
