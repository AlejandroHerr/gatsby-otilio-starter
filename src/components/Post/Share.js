import React from 'react';
import PropTypes from 'prop-types';
import { FacebookShareButton } from 'react-share';
import SimpleIcon from '../icons/SimpleIcon';

import styles from './Post.module.scss';

const Share = ({ path, url }) => (
  <div className={styles.post_share}>
    <FacebookShareButton url={`${url}${path}`} className={styles.post_share__button}>
      <SimpleIcon brand="facebook" className={styles.post_share__facebook} />
    </FacebookShareButton>
  </div>
);

Share.defaultProps = {
  path: '',
};

Share.propTypes = {
  path: PropTypes.string,
  url: PropTypes.string.isRequired,
};

export default Share;
