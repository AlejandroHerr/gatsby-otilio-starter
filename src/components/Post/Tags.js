// @flow
import React from 'react';

import PostTag from './Tag';

import styles from './Post.module.scss';

const Tags = ({ tags }: { tags: Array<string>}) => (
  <aside className={styles.post_tags}>
    {tags.map(tag => <PostTag key={tag} tag={tag} />)}
  </aside>
);

export default Tags;
