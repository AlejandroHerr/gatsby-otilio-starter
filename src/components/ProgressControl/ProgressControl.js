// @flow
/* global window,document */
import React, { PureComponent, type Node } from 'react';

import ProgressControlBar from './ProgressControlBar';

type PropsType = {
  children: Node,
  speed: number,
};

type StateType = {
  finished: boolean,
  progress: number,
};

export const defaultMapPropsToStyle = ({ progress, speed }: {progress: number, speed: number}) => ({
  width: `${progress}%`,
  transition: `transform ${speed}ms ease`,
});

export default class ProgressControl extends PureComponent<PropsType, StateType> {
  container: ?HTMLDivElement;
  calculateProgress: () => void;

  static defaultProps: {
    speed: number,
  }

  constructor(props: PropsType) {
    super(props);

    this.state = {
      finished: false,
      progress: 0,
    };

    this.container = null;
    this.calculateProgress = this.calculateProgress.bind(this);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.calculateProgress);
    window.addEventListener('resize', this.calculateProgress);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.calculateProgress);
    window.removeEventListener('resize', this.calculateProgress);
  }

  calculateProgress() {
    if (this.container) {
      const { innerHeight, pageYOffset } = window;
      const { clientHeight, offsetTop } = this.container;

      const progress = (100 * pageYOffset) / ((offsetTop + clientHeight) - innerHeight);

      this.setState({ finished: progress > 100, progress: Math.min(progress, 100) });
    }
  }

  render() {
    const { children, speed } = this.props;
    const { finished, progress } = this.state;

    return (
      <div ref={(ref) => { this.container = ref; }}>
        <ProgressControlBar
          finished={finished}
          progress={progress}
          speed={speed}
          mapPropsToStyle={defaultMapPropsToStyle}
        />
        {children}
      </div>
    );
  }
}

ProgressControl.defaultProps = {
  speed: 200,
};
