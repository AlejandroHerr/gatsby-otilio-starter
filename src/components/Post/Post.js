/* eslint-disable react/no-danger */
import React from 'react';
import PropTypes from 'prop-types';

import {
  authorType,
  postPreviewType,
  postType,
  siteInfoType,
} from '../../propTypes/post';

import ProgressControl from '../ProgressControl';

import PostHeader from './PostHeader';
import Tags from './Tags';
import Share from './Share';
import Author from './Author';
import Navigation from './Navigation';

import styles from './Post.module.scss';

const Post = ({
  author, navigation, post, siteInfo,
}) => {
  const { next, prev } = navigation;
  const {
    slug: path,
  } = post.fields;
  const {
    author: postAuthor, title, cover, tags, date,
  } = post.frontmatter;

  const { url } = siteInfo;


  const coverImage = cover && (cover.childImageSharp.resolutions || cover.childImageSharp.sizes);
  return (
    <div>
      <PostHeader
        author={postAuthor}
        date={date}
        title={title}
        image={coverImage}
        timeToRead={post.timeToRead}
      />
      <main>
        <article>
          <div className={styles.post__inner}>
            <section className={styles.post__content}>
              <ProgressControl>
                <div dangerouslySetInnerHTML={{ __html: post.html }} />
              </ProgressControl>
            </section>
            <section className={styles.post__info}>
              <Share path={path} url={url} />
              {tags && tags.length && (<Tags tags={tags} />)}
              <div className={styles.post__clear} />
              <Author author={author} />
              <div className="clear" />
            </section>
            <Navigation next={next} prev={prev} />
          </div>
        </article>
      </main>
    </div>
  );
};

Post.propTypes = {
  author: authorType.isRequired,
  navigation: PropTypes.shape({
    next: postPreviewType,
    prev: postPreviewType,
  }).isRequired,
  post: postType.isRequired,
  siteInfo: siteInfoType.isRequired,
};


export default Post;
