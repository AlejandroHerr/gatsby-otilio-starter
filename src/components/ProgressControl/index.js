// @flow
import ProgressControl from './ProgressControl';

export default ProgressControl;
