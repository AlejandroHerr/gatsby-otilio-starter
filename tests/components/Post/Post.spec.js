import React from 'react';
import { shallow } from 'enzyme';

import Post from '../../../src/components/Post/';
import PostHeader from '../../../src/components/Post/PostHeader';
import ProgressControl from '../../../src/components/ProgressControl';
import Share from '../../../src/components/Post/Share';
import Tags from '../../../src/components/Post/Tags';
import Author from '../../../src/components/Post/Author';
import Navigation from '../../../src/components/Post/Navigation';

const defaultProps = {
  siteInfo: {
    url: 'siteInfo.url',
  },
  post: {
    fields: {
      slug: 'slug',
    },
    frontmatter: {
      author: 'autor',
      title: 'title',
      date: '1',
    },
  },
  author: {},
  navigation: {},
};

const setup = (props = {}) => {
  const post = shallow(<Post {...defaultProps} {...props} />);

  return {
    post,
    postHeader: post.find(PostHeader),
    progressControl: post.find(ProgressControl),
    share: post.find(Share),
    tags: post.find(Tags),
    author: post.find(Author),
    navigation: post.find(Navigation),
  };
};

describe('Post', () => {
  it('should render a Post', () => {
    const {
      post, postHeader, progressControl, share, tags, author, navigation,
    } = setup();

    expect(postHeader).toHaveLength(1);
    expect(progressControl).toHaveLength(1);
    expect(share).toHaveLength(1);
    expect(tags).toHaveLength(0);
    expect(author).toHaveLength(1);
    expect(navigation).toHaveLength(1);
    expect(post).toMatchSnapshot();
  });
  it('should render a Post with tags', () => {
    const postProp = {
      ...defaultProps.post,
      frontmatter: {
        ...defaultProps.post.frontmatter,
        tags: ['test1', 'test2'],
      },
    };
    const {
      post, postHeader, progressControl, share, tags, author, navigation,
    } = setup({ post: postProp });

    expect(postHeader).toHaveLength(1);
    expect(progressControl).toHaveLength(1);
    expect(share).toHaveLength(1);
    expect(tags).toHaveLength(1);
    expect(author).toHaveLength(1);
    expect(navigation).toHaveLength(1);
    expect(post).toMatchSnapshot();
  });
  it('should render a PostHeader with cover', () => {
    const postProp = {
      ...defaultProps.post,
      frontmatter: {
        ...defaultProps.post.frontmatter,
        cover: {
          childImageSharp: {
            sizes: { src: 'src' },
          },
        },
      },
    };
    const {
      post, postHeader,
    } = setup({ post: postProp });

    expect(postHeader.prop('image')).toBe(postProp.frontmatter.cover.childImageSharp.sizes);
    expect(post).toMatchSnapshot();
  });
});
