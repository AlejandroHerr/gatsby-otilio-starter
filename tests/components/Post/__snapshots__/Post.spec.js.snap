// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should render a Post 1`] = `
<div>
  <PostHeader
    author="autor"
    date="1"
    image={null}
    title="title"
  />
  <main>
    <article>
      <div
        className="post__inner"
      >
        <section
          className="post__content"
        >
          <ProgressControl
            speed={200}
          >
            <div
              dangerouslySetInnerHTML={
                Object {
                  "__html": undefined,
                }
              }
            />
          </ProgressControl>
        </section>
        <section
          className="post__info"
        >
          <Share
            path="slug"
            url="siteInfo.url"
          />
          <div
            className="post__clear"
          />
          <Author
            author={Object {}}
          />
          <div
            className="clear"
          />
        </section>
        <Navigation
          next={null}
          prev={null}
        />
      </div>
    </article>
  </main>
</div>
`;

exports[`Post should render a Post with tags 1`] = `
<div>
  <PostHeader
    author="autor"
    date="1"
    image={null}
    title="title"
  />
  <main>
    <article>
      <div
        className="post__inner"
      >
        <section
          className="post__content"
        >
          <ProgressControl
            speed={200}
          >
            <div
              dangerouslySetInnerHTML={
                Object {
                  "__html": undefined,
                }
              }
            />
          </ProgressControl>
        </section>
        <section
          className="post__info"
        >
          <Share
            path="slug"
            url="siteInfo.url"
          />
          <Tags
            tags={
              Array [
                "test1",
                "test2",
              ]
            }
          />
          <div
            className="post__clear"
          />
          <Author
            author={Object {}}
          />
          <div
            className="clear"
          />
        </section>
        <Navigation
          next={null}
          prev={null}
        />
      </div>
    </article>
  </main>
</div>
`;

exports[`Post should render a PostHeader with cover 1`] = `
<div>
  <PostHeader
    author="autor"
    date="1"
    image={
      Object {
        "src": "src",
      }
    }
    title="title"
  />
  <main>
    <article>
      <div
        className="post__inner"
      >
        <section
          className="post__content"
        >
          <ProgressControl
            speed={200}
          >
            <div
              dangerouslySetInnerHTML={
                Object {
                  "__html": undefined,
                }
              }
            />
          </ProgressControl>
        </section>
        <section
          className="post__info"
        >
          <Share
            path="slug"
            url="siteInfo.url"
          />
          <div
            className="post__clear"
          />
          <Author
            author={Object {}}
          />
          <div
            className="clear"
          />
        </section>
        <Navigation
          next={null}
          prev={null}
        />
      </div>
    </article>
  </main>
</div>
`;
